### ISSUE TITLE
A more robust manifest format

### ISSUE THREAD

**Issue by rekka** [tectonic-typesetting/tectonic#74]
Currently the manifest format for the local cache is stored in a space-separated format without any escaping. This leads to invalid entries being inserted if for instance files with spaces are requested:

```tex
\input{"with space.tex"}\bye
```

produces entries

```
with space.tex 0 -
with space.tex.tex 0 -
```

It is not really a serious issue at the moment, but a future-proofing of the format is desirable.

I believe that one appropriate solution would be to use the [csv crate](https://crates.io/crates/csv), which also allows for [space-separated entries](https://docs.rs/csv/1.0.0-beta.3/csv/struct.ReaderBuilder.html#method.delimiter). So it should be backwards compatible with the current format. It also offers easy parsing using `serde`. Moreover, it will  most likely hit version 1 soon.

I am happy to implement this.

**Comment by pkgw** [comment]
Yes, it would be good to tighten things up here.

I might prefer to stick with the current format but just reverse the ordering: size, SHA256, filename. That way you can just split on the first two spaces and take the rest as a lump.

Although I guess we're still fragile if anyone uses a filename containing a newline ...

**Comment by rekka** [comment]
That is actually a pretty good solution too. Now files with newlines...

**Comment by alexander-bauer** [comment]
I was just browsing through looking at this, and poked around until I found the relevant parts of the code. It looks like [`src/io/local_cache:176`](https://github.com/tectonic-typesetting/tectonic/blob/master/src/io/local_cache.rs#L176) is where it gets written; I'm not sure where it gets read from the manifest.

**Comment by rekka** [comment]
It is a bit above that, [in the `new` method](https://github.com/tectonic-typesetting/tectonic/blob/c85e179e3f22a07d9c01866b3cc807ce05dd3a04/src/io/local_cache.rs#L114-L122).

**Comment by alexander-bauer** [comment]
@rekka I can go ahead and make a PR for this, but I'd also hate to duplicate work, if you'd like to

**Comment by rekka** [comment]
I'm not currently working on this, but I think we should reach a consensus with @pkgw on what the best approach here is.

I am not leaning towards a simple format as suggested by @pkgw with: size SHA256 filename, and completely disallowing files with new lines to enter the manifest. This can be easily parsed by `str::splitn` method.

**Comment by pkgw** [comment]
@rekka Do you perhaps mean "now" and not "not" in your most recent comment?

I think this is a good change but I am a little bit concerned about how to transition from an old manifest format to a new one. I guess we can just munge a version code into the file name, which seems reasonably future-proof if we want to make any other changes to the format.

**Comment by rekka** [comment]
Ah, sorry about that, I meant "not", not working on this. 

A version code is a good option. I suppose you mean that tectonic would just discard an outdated manifest and rebuild the cache from scratch. That's simple.

**Comment by pkgw** [comment]
Yes, re-downloading files is a bit annoying but it seems like the lowest-pain option. I don't want people to have to be told to manually delete their cache directories if I can avoid it.

**Comment by alexander-bauer** [comment]
An option that's a little less rough on the UX would be to migrate from the older manifest version automatically, but I suspect that maintaining the code for that (and all future manifest improvements) would be daunting.

Perhaps the UX problem could be offset (if it needs to be at all) by printing an explanatory message when creating the new manifest if an old one is already present?
